cpp/oneapi/dal/detail/mpi/communicator.hpp

```diff
@@ -0,0 +1,314 @@
+            rank_count_ = rank_count;
+        }
+        return rank_count_;
+    }
+
+    std::int64_t get_default_root_rank() override {
+        return default_root_;
+    }
+
+    void barrier() override {
+        mpi_call(MPI_Barrier(mpi_comm_));
+    }
+
+    spmd_request_iface* bcast(byte_t* send_buf,
+                              std::int64_t count,
+                              const data_type& dtype,
+                              std::int64_t root) override {
+        ONEDAL_ASSERT(root >= 0);
+
+        if (count == 0) {
+            return nullptr;
+        }
+
+        ONEDAL_ASSERT(send_buf);
+        ONEDAL_ASSERT(count > 0);
+
+        MPI_Request mpi_request;
+        mpi_call(MPI_Ibcast(send_buf,
+                            integral_cast<int>(count),
+                            make_mpi_data_type(dtype),
+                            integral_cast<int>(root),
+                            mpi_comm_,
+                            &mpi_request));
+
+        return new mpi_request_impl{ mpi_request };
+    }
+
+    spmd_request_iface* gather(const byte_t* send_buf,
+                               std::int64_t send_count,
+                               byte_t* recv_buf,
+                               std::int64_t recv_count,
+                               const data_type& dtype,
+                               std::int64_t root) override {
+        ONEDAL_ASSERT(root >= 0);
+
+        if (send_count == 0) {
+            return nullptr;
+        }
+
+        ONEDAL_ASSERT(send_buf);
+        if (get_rank() == root) {
+            ONEDAL_ASSERT(recv_buf);
+        }
+
+        MPI_Request mpi_request;
+        mpi_call(MPI_Igather(send_buf,
+                             integral_cast<int>(send_count),
+                             make_mpi_data_type(dtype),
+                             recv_buf,
+                             integral_cast<int>(recv_count),
+                             make_mpi_data_type(dtype),
+                             integral_cast<int>(root),
+                             mpi_comm_,
+                             &mpi_request));
+
+        return new mpi_request_impl{ mpi_request };
+    }
+
+    spmd_request_iface* gatherv(const byte_t* send_buf,
+                                std::int64_t send_count,
+                                byte_t* recv_buf,
+                                const std::int64_t* recv_counts,
+                                const std::int64_t* displs,
+                                const data_type& dtype,
+                                std::int64_t root) override {
+        ONEDAL_ASSERT(root >= 0);
+
+        if (send_count == 0) {
+            return nullptr;
+        }
+
+        ONEDAL_ASSERT(send_buf);
+        if (get_rank() == root) {
+            ONEDAL_ASSERT(recv_counts);
+            ONEDAL_ASSERT(displs);
+            ONEDAL_ASSERT(recv_buf);
+        }
+
+        array<int> recv_counts_int;
+        array<int> displs_int;
```

Comment:
Please don't use `int`. Can you use `std::int32_t` instead?

Reply:
Should be int for MPI